### PR TITLE
dockerTools: fixes extraCommands for mkRootLayer.

### LINKED ISSUE #34779: dockerTools.buildImageWithNixDb fails when also using runAsRoot
## Issue description

When using `runAsRoot` with `buildImageWithNixDb` using a recent-enough nixpkgs checkout, it is failing with:

```
/nix/store/...-stdenv/setup: eval: line ____: syntax error near unexpected token `('
```

The exact line changes depending on the script built by `dockerTools`, which depends on what is added to `runAsRoot`.

### Steps to reproduce

Use cb6fc52f998f8f3106a1506c052cffae752e8b7c (or later up to today) and try producing an image using both `runAsRoot` and `buildImageWithNixDb`.

`repro.nix`

```nix
let
  hostPkgs = import <nixpkgs> {};
  pinnedPkgs = hostPkgs.fetchFromGitHub {
    owner = "NixOS";
    repo = "nixpkgs-channels";
    # First revision failing.
    #rev = "cb6fc52f998f8f3106a1506c052cffae752e8b7c";
    #sha256 = "0fv25qj85qn9rhrlh7kbgddjy99dh96jrdw0vk08l09g3qk0x8fi";
    # Last revision working.
    rev = "df589a438e01c06a6661cbd18bfe8f568d8b7f83";
    sha256 = "1y75nw9n2rjz925hd21bs91azaggm4ry2h886hnd0pjkrq07h8gg";
  };
  pkgs = import pinnedPkgs { config = {}; };
in
  with pkgs;
  dockerTools.buildImageWithNixDb {
    name = "nix";
    contents = [
      coreutils
      nix
    ];
    runAsRoot = ''
      #!${pkgs.stdenv.shell}
      echo "everything is working fine"
    '';
    config = {
      Env = [ "NIX_PAGER=cat" ];
    };
  }
```

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

Here they are, but should be irrelevant AFAIK since it's using a pinned nixpkgs checkout.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.15, NixOS, 17.09.git.c1d9aff (Hummingbird)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 1.11.16`
 - channels(root): `"nixos-17.09.2955.76614894f34"`
 - nixpkgs: `/etc/nixos/nixpkgs`
```

The issue, while being present from that commit onward, is possibly not *caused* by that commit. As far as I understand it from a shallow inspection, it seems to come from the way the scripts are ran with dockerTools, it is an `eval` of variables. Something spooky is possibly happening when the script is `????` into the qemu vm for use with `runAsRoot`. The failure, in some tests I did, comes also(?) from this line of comments:

```
        echo "Copying everything to /nix/store (will take a while)..."
```

It would be the parenthesis present in that line which is causing an issue.

### PR DESCRIPTION
**Pull request by samueldr**
The extraCommands was, previously, simply put in the body of the script
using nix expansion `${extraCommands}` (which looks exactly like bash
expansion!).

This causes issues like in #34779 where scripts will eventually create
invalid bash.

The solution used was to use `$extraCommands` the way `mkPureLayer`
does, by using the variable from the environment instead of expanding
it inside the script.

 * * *

Fixes #34779

###### Motivation for this change

See #34779, this fixes it.

###### Things done

- ✔️ Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - ✔️ NixOS
   - 🆖 macOS
   - ❌ other Linux distributions
- ✔️ Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- ❌ Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- 🆖 Tested execution of all binary files (usually in `./result/bin/`)
- ✔️ Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This was verified using this:

`nix-build -I nixpkgs=/path/to/checkout/of/nixpkgs repro.nix`

Using this `repro.nix`:
```nix
with import <nixpkgs> {};
  dockerTools.buildImageWithNixDb {
    name = "nix";
    contents = [
      coreutils
      nix
    ];
    runAsRoot = ''
      #!${pkgs.stdenv.shell}
      echo "everything is working fine"
    '';
    config = {
      Env = [ "NIX_PAGER=cat" ];
    };
  }
```

This is now verified in a nixos test

```
NIX_PATH=nixpkgs=/path/to/checkout/of/nixpkgs nix-build '<nixpkgs/nixos/tests/docker-tools.nix>'
```

Apply the test before the fix to verify the test tests for regression.

---

*cc @nlewo (who seemed to work on the docker tooling a bunch)*

### CODE REVIEW

**Comment by nlewo** [comment]
Thanks for your investigation!

I think your patch is not working since the `extraCommands` would not be present in the `mkRootLayer` build environment: the image is built because `extraCommands` are not evaluated :)
This is why a `run-as-root` script is created in the `mkRootLayer` function. So you should also create such kind of script for the `extraCommands` attribute.

Regarding the regression test, you could also reproduce with this expression (which is maybe more close to the issue since is not related to `buildImageWithNixDb`)
```
runAsRootExtraCommands = buildImage {
    name = "runAsRootExtraCommands";
    contents = [ pkgs.coreutils ];
    runAsRoot = ''echo "(runAsRoot)" > runAsRoot'';
    extraCommands = ''echo "extraCommand" > extraCommands'';
  };
```
At run time, we could check these files exist:
```
docker run runasrootextracommands cat extraCommands
docker run runasrootextracommands cat runAsRoot
```

**Comment by samueldr** [comment]
*(Put as WIP as I will verify the images later)*

**Comment by samueldr** [comment]
Oh wow, thanks for the tip on how to write a more proper test. I'm not used to docker, so I didn't know what and how to test. It's an elegant and simple way to check if the commands ran.

I'm not sure whether the `unshare` bit from `runAsRoot` is needed. I think it would not since the previous embedding didn't do anything like it.

I have, once again, checked the new test with and without the change. It does fail before my new fix, and does pass after the new fix. I tested it also with the old fix, and it failed, so you were right!

The PR has been updated accordingly.

File: nixos/tests/docker-tools.nix [hunk 1]
```diff
@@ -16,6 +16,16 @@ import ./make-test.nix ({ pkgs, ... }: {
       };
   };
 
+  runAsRootExtraCommands = pkgs.dockerTools.buildImage {
+    name = "runAsRootExtraCommands";
+    contents = [ pkgs.coreutils ];
+    # The parens here are to create problematic bash to embed and eval. In case
+    # this is *embedded* into the script (with nix expansion) the initial quotes
+    # will close the string and the following parens are unexpected
```
Comment (nlewo) [inline nixos/tests/docker-tools.nix]: You can also mention the issue you've opened

File: pkgs/build-support/docker/default.nix [hunk 2]
```diff
@@ -360,7 +360,9 @@ rec {
     extraCommands ? ""
   }:
     # Generate an executable script from the `runAsRoot` text.
-    let runAsRootScript = shellScript "run-as-root.sh" runAsRoot;
+    let
+      runAsRootScript = shellScript "run-as-root.sh" runAsRoot;
+      extraCommandsScript = shellScript "extracommands.sh" extraCommands;
```
Comment (nlewo) [inline pkgs/build-support/docker/default.nix]: Could you call it `extra-commands.sh` since `run-as-root.sh` is "dashed"

**Comment by nlewo** [comment]
Thanks for you changes. There are just some minor comments.
I also think we don't have to use `unshare` for `extraCommands`.

**Comment by samueldr** [comment]
*Updated with the relevant changes.*

File: nixos/tests/docker-tools.nix [hunk 3]
```diff
@@ -16,6 +16,17 @@ import ./make-test.nix ({ pkgs, ... }: {
       };
   };
 
+  runAsRootExtraCommands = pkgs.dockerTools.buildImage {
```
Comment (nlewo) [inline nixos/tests/docker-tools.nix]: Arf, sorry, I haven't seen this image was in this file:/ Could you either move it to `docker/examples.nix` or in a 
```nix
let runAsRootExtraCommands =... 
in import ./make-test.nix....
```

**Comment by nlewo** [comment]
@GrahamcOfBorg test docker-tools

**Comment by nlewo** [comment]
Thanks!

**Pull request merged**